File: two-sum.java
import java.util.HashMap;

public class Solution {
    public int[] twoSum(int[] numbers, int target) {
        HashMap<Integer, Integer> numIndex = new HashMap<Integer, Integer>();
        for (int i = 0; i < numbers.length; i++) {
            Integer need_index = numIndex.get(target - numbers[i]);
            if (need_index != null) {
                int[] ans = {need_index + 1, i + 1};
                return ans;
            }
            numIndex.put(numbers[i], i);
        }
        return null;
    }
}

File: two-sum.cc
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int> &numbers, int target) {
    typedef unordered_map<int, int> NumberIndex;
    unordered_map<int, int> num_pos;
    for (int i = 0; i < numbers.size(); i++) {
      NumberIndex::iterator it = num_pos.find(target - numbers[i]);
      if (it != num_pos.end()) {
        vector<int> ans;
        ans.push_back(it->second + 1);
        ans.push_back(i + 1);
        return ans;
      }
      num_pos[numbers[i]] = i;
    }
  }
};
